feat(output plugin): Add new output plugin to support Apache IoTDB (#11557)
influxdata/telegraf: plugins/outputs/iotdb/README.md

# IoTDB Output Plugin

This output plugin saves Telegraf metrics to an Apache IoTDB backend,
supporting session connection and data insertion.

## Apache IoTDB

Apache IoTDB (Database for Internet of Things) is an IoT native database with
high performance for data management and analysis, deployable on the edge and
the cloud. Due to its light-weight architecture, high performance and rich
feature set together with its deep integration with Apache Hadoop, Spark and
Flink, Apache IoTDB can meet the requirements of massive data storage,
high-speed data ingestion and complex data analysis in the IoT industrial
fields.

For more details consult the [Apache IoTDB website](https://iotdb.apache.org)
or the [Apache IoTDB GitHub page](https://github.com/apache/iotdb).

## Getting started

Before using this plugin, please configure the IP address, port number,
user name, password and other information of the database server,
as well as some data type conversion, time unit and other configurations.

Please see the [configuration section](#Configuration) for an example
configuration.

> ## Please pay attention to these points
> 
> 1. IoTDB currently **DOES NOT support unsigned integer**.
> There are three available options of converting uint64, which are specified by
> parameter `uint64_conversion`.
> 
>    - `int64_clip`, default option. If an unsigned integer is greater than
>    `math.MaxInt64`, save it as `int64`; else save `math.MaxInt64`
>    (9223372036854775807).
>    - `int64`, force converting an unsigned integer to a`int64`,no mater
>    what the value it is. This option may lead to exception if the value is
>    greater than `int64`.
>    - `text`force converting an unsigned integer to a string, no mater what the
>    value it is.
> 
> 2. IoTDB supports a variety of time precision. You can specify which precision
> you want using the `timestamp_precision` option. Default is `nanosecond`.
> Other options are "second", "millisecond", "microsecond".
> 
> 3. Till now, IoTDB can not support Tag indexing well. To see current process
>    method, please refer to [InfluxDB-Protocol Adapter](
>    https://iotdb.apache.org/UserGuide/Master/API/InfluxDB-Protocol.html).
>    There are two available options of converting tags, which are specified by
>    parameter `convert_tags_to`:
> 
>    - `fields`. Treat Tags as measurements. For each Key:Value in Tag,
>    convert them into Measurement, Value, DataType, which are supported in IoTDB.
>    - `device_id`, default option. Treat Tags as part of device id. Tags
>    is subtree of 'Name'.
> 
>    For example, there is a metric:
> 
>    `Name="root.sg.device", Tags={tag1="private", tag2="working"}, Fields={s1=100, s2="hello"}`
> 
>    - `fields`, result: `root.sg.device, s1=100, s2="hello", tag1="private", tag2="working"`
>    - `device_id`, result: `root.sg.device.private.working, s1=100, s2="hello"`

Reviewer: Let's use the heading here to describe what it contains. (We would also hope people are paying attention to the whole document, not just these points :smiley:)

The items in the ordered list are large enough that they would work better as subheadings.

You use the terms "parameter" or "option" here but telegraf docs call it "setting".

```suggestion
## Metric Translation
IoTDB uses a different data format for metric data than telegraf. It is important to note that depending on the metrics being written, the translation may be lossy. This plugin translates to IoTDB format in the following ways:

### Unsigned Integers
IoTDB currently **DOES NOT support unsigned integer**.
There are three available options of converting uint64, which are specified by
setting `uint64_conversion`.

   - `int64_clip`, default option. If an unsigned integer is greater than
   `math.MaxInt64`, save it as `int64`; else save `math.MaxInt64`
   (9223372036854775807).
   - `int64`, force converting an unsigned integer to a`int64`,no mater
   what the value it is. This option may lead to exception if the value is
   greater than `int64`.
   - `text`force converting an unsigned integer to a string, no mater what the
   value it is.

### Time Precision
IoTDB supports a variety of time precision. You can specify which precision
you want using the `timestamp_precision` setting. Default is `nanosecond`.
Other options are "second", "millisecond", "microsecond".

### Metadata (tags)
IoTDB uses a tree model for metadata while Telegraf uses a tag model.
(See [InfluxDB-Protocol Adapter](
https://iotdb.apache.org/UserGuide/Master/API/InfluxDB-Protocol.html)
There are two available options of converting tags, which are specified by
setting `convert_tags_to`:

   - `fields`. Treat Tags as measurements. For each Key:Value in Tag,
   convert them into Measurement, Value, DataType, which are supported in IoTDB.
   - `device_id`, default option. Treat Tags as part of device id. Tags
   is subtree of 'Name'.

   For example, there is a metric:

   `Name="root.sg.device", Tags={tag1="private", tag2="working"}, Fields={s1=100, s2="hello"}`

   - `fields`, result: `root.sg.device, s1=100, s2="hello", tag1="private", tag2="working"`
   - `device_id`, result: `root.sg.device.private.working, s1=100, s2="hello"`
```

Author: Thanks for suggestions! `README.md` updated. Relevant comments in `sample.conf` are also updated.

This output plugin saves Telegraf metrics to an Apache IoTDB backend,
supporting session connection and data insertion.

## Apache IoTDB

Apache IoTDB (Database for Internet of Things) is an IoT native database with
high performance for data management and analysis, deployable on the edge and
the cloud. Due to its light-weight architecture, high performance and rich
feature set together with its deep integration with Apache Hadoop, Spark and
Flink, Apache IoTDB can meet the requirements of massive data storage,
high-speed data ingestion and complex data analysis in the IoT industrial
fields.

For more details consult the [Apache IoTDB website](https://iotdb.apache.org)
or the [Apache IoTDB GitHub page](https://github.com/apache/iotdb).

## Getting started

Before using this plugin, please configure the IP address, port number,
user name, password and other information of the database server,
as well as some data type conversion, time unit and other configurations.

Please see the [configuration section](#Configuration) for an example
configuration.

## Please pay attention to these points

1. IoTDB (version 0.13.x or older) **DO NOT support unsigned integer**.

> There are three available options of converting uint64, which are specified by

Author: All right! Thanks for your suggestion.

you want using the `timestamp_precision` option. Default is `nanosecond`.
Other options are "second", "millisecond", "microsecond".

3. Till now, IoTDB can not support Tag indexing well. To see current process
   method, please refer to [InfluxDB-Protocol Adapter](
   https://iotdb.apache.org/UserGuide/Master/API/InfluxDB-Protocol.html).
   There are two available options of converting tags, which are specified by
   parameter `convert_tags_to`:

   - `fields`. Treat Tags as measurements. For each Key:Value in Tag,
   convert them into Measurement, Value, DataType, which are supported in IoTDB.
   - `device_id`, default option. Treat Tags as part of device id. Tags
   is subtree of 'Name'.

   For example, there is a metric:

   `Name="root.sg.device", Tags={tag1="private", tag2="working"}, Fields={s1=100, s2="hello"}`

   - `fields`, result: `root.sg.device, s1=100, s2="hello", tag1="private", tag2="working"`
   - `device_id`, result: `root.sg.device.private.working, s1=100, s2="hello"`

## Testing

Testing contains: network connection, error correction, datatype conversion,
data writing.

Use `-short` option to make testing shorter. This mode skips integration tests.

> 

Author: ok. Section removed. Thanks.

## Configuration

```toml @sample.conf
# Save metrics to an IoTDB Database
[[outputs.iotdb]]
  ## Configuration of IoTDB server connection
  host = "127.0.0.1"

>   # port = "6667"

Author: I' sorry. It's my bad that I did not run `make docs`. I will run it after next commit.

I guess it will automatically correct my `readme` and attach `sample.conf`, right? However, it did not worked, perhaps because of the local settings of windows. I need to manually modify the `makefile` to make it run, but after running the command, nothing really happened. 

But I have made sure that the contents in `readmd` are same with  `sample.conf`.

Thanks

>   # port = "6667"

Reviewer 2: It worked, the REAME is adapted... Yes, the `@sample.conf` statement in the toml section will indicate that you want to embed that file at this location to the tool I wrote (`tools/readme_config_includer`). Let me know (i.e. create an issue) if you have trouble on Windows!

high-speed data ingestion and complex data analysis in the IoT industrial
fields.

For more details consult the [Apache IoTDB website](https://iotdb.apache.org)
or the [Apache IoTDB GitHub page](https://github.com/apache/iotdb).

## Getting started

Before using this plugin, please configure the IP address, port number,
user name, password and other information of the database server,
as well as some data type conversion, time unit and other configurations.

Please see the [configuration section](#Configuration) for an example
configuration.

## Please pay attention to these points

1. IoTDB (version 0.13.x or older) **DO NOT support unsigned integer**.
There are three available options of converting uint64, which are specified by
parameter `uint64_conversion`.

   - `int64_clip`, default option. If an unsigned integer is greater than
   `math.MaxInt64`, save it as `int64`; else save `math.MaxInt64`
   (9223372036854775807).
   - `int64`, force converting an unsigned integer to a`int64`,no mater
   what the value it is. This option may lead to exception if the value is
   greater than `int64`.
   - `text`force converting an unsigned integer to a string, no mater what the

>    value it is.

Reviewer 2: Just for understanding. Do newer versions support unsigned integers? If so, we should add an option `"none"` that does not do any conversion for those newer version and make `"none"` the default...

Author: This is really a good idea! I haven't thought about such a thing before.
But at the meeting, the team of Apache IoTDB said that it was not necessary to support unsigned int. The team is currently improving distributed service structure, and unsigned int problem will not be considered in the next few versions. 
Therefore, I think current options is ok. :-)

> 2. IoTDB supports a variety of time precision, but no matter what precision,
> timestamp is stored in `Int64`, so users need to specify the unit of timestamp.
> Default unit is `nanosecond`.
> 

Reviewer 2: Maybe
```suggestion
2. IoTDB supports a variety of time precision. You can specify which precision you want using the `timestamp_precision` option. Default is `nanosecond`.

```

Author: Great! Non-programmers really don't need to care about these details.

3. Till now, IoTDB can not support Tag indexing well. To see current process
   method, please refer to [InfluxDB-Protocol Adapter](
   https://iotdb.apache.org/UserGuide/Master/API/InfluxDB-Protocol.html).
   There are two available options of converting tags, which are specified by
   parameter `convert_tags_to`:

   - `fields`. Treat Tags as measurements. For each Key:Value in Tag,
   convert them into Measurement, Value, DataType, which are supported in IoTDB.
   - `device_id`, default option. Treat Tags as part of device id. Tags
   is subtree of 'Name'.

   For example, there is a metric:

   `Name="root.sg.device", Tags={tag1="private", tag2="working"}, Fields={s1=100, s2="hello"}`

   - `fields`, result: `root.sg.device, s1=100, s2="hello", tag1="private", tag2="working"`
   - `device_id`, result: `root.sg.device.private.working, s1=100, s2="hello"`

## Testing

**Please prepare running database before testing**.
Target address is`localhost:6667` by default, which can be edit in
`iotdb_test.go`. `test_host` is the target ip address of database server.

Testing contains: network connection, error correction, datatype conversion,

> data writing.

Reviewer 2: I think this can be removed when switching to `testcontainers`!?

Author: Oh yes! At last commit, `testcontainers` was not added. So, this is an outdated description. Now it's fixed. Thanks.

# IoTDB Output Plugin

We provide readme and sample configuration file in different languages.
Please Choose a friendly one here:
[English](./README.md) | [中文](./README_ZH.md)

> 

Author: All right, I ‘m happy to accept this suggestion to make the documentation meet Telegraf standards. But the reason why I did this is that Apache IoTDB fully supports Chinese documents. I will remove all non-English versions at next commit.

> 

Reviewer 2: Feel free to host this elsewhere and we link to it, but please understand that we are not able to maintain multiple languages. Think of the case someone adds a new option, who will document it in Chinese or German or French or ...?

The IoTDB output plugin saves Telegraf metric data to an IoTDB backend.
This plugin uses Apache IoTDB client for Golang to support session connection
and data insertion. For metrics sampled, this pulgin converts them into records

> and inserts them into databases.

Author: This is really a great suggestion. The session is more concise, and users don't need to care about the implementation details. I will adopt it, thanks.

## Apache IoTDB

Apache IoTDB (Database for Internet of Things) is an IoT native database with

> high performance for data management and analysis, deployable on the edge and

Author: Ok, it's my bad to put links in an inconspicuous place. I will fix it.

> high performance for data management and analysis, deployable on the edge and

Reviewer 2: Yeah saw it after my comment, but I think with your change it is easier to find.

the cloud. Due to its light-weight architecture, high performance and rich
feature set together with its deep integration with Apache Hadoop, Spark and
Flink, Apache IoTDB can meet the requirements of massive data storage,
high-speed data ingestion and complex data analysis in the IoT industrial
fields.

Apache IoTDB website: <https://iotdb.apache.org>

> Apache IoTDB Github: <https://github.com/apache/iotdb>

Author: This is indeed a better expression, thanks for suggestion.

## Apache IoTDB Client for Golang

There is an Apache IoTDB Client for Golang, using native Golang API to
inserting, deleting, modifying records.

> Apache IoTDB Golang Client Github: <https://github.com/apache/iotdb>

Author: All right.

## Getting started

Before using this plugin, please configure the IP address, port number,
user name, password and other information of the database server,
as well as some data type conversion, time unit and other configurations.

There is a sample configuration: [English Configuration](./sample.conf).
And there is one in Chinese:  [中文配置样例](./sample_zh.conf).

> The configuration is also provided at the end of this readme.

Author: It's really better, thanks.
